src/Workspaces/Core/Portable/Execution/Extensions.cs

```diff
@@ -15,106 +15,30 @@ public static T[] ReadArray<T>(this ObjectReader reader)
-            }
-
-            if (value is ProjectStateChecksums)
-            {
-                return WellKnownSynchronizationKinds.ProjectState;
-            }
-
-            if (value is DocumentStateChecksums)
-            {
-                return WellKnownSynchronizationKinds.DocumentState;
-            }
-
-            if (value is ProjectChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.Projects;
-            }
-
-            if (value is DocumentChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.Documents;
-            }
-
-            if (value is TextDocumentChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.TextDocuments;
-            }
-
-            if (value is ProjectReferenceChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.ProjectReferences;
-            }
-
-            if (value is MetadataReferenceChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.MetadataReferences;
-            }
-
-            if (value is AnalyzerReferenceChecksumCollection)
-            {
-                return WellKnownSynchronizationKinds.AnalyzerReferences;
-            }
-
-            if (value is SolutionInfo.SolutionAttributes)
-            {
-                return WellKnownSynchronizationKinds.SolutionAttributes;
-            }
-
-            if (value is ProjectInfo.ProjectAttributes)
-            {
-                return WellKnownSynchronizationKinds.ProjectAttributes;
-            }
-
-            if (value is DocumentInfo.DocumentAttributes)
-            {
-                return WellKnownSynchronizationKinds.DocumentAttributes;
-            }
-
-            if (value is CompilationOptions)
-            {
-                return WellKnownSynchronizationKinds.CompilationOptions;
-            }
-
-            if (value is ParseOptions)
-            {
-                return WellKnownSynchronizationKinds.ParseOptions;
-            }
-
-            if (value is ProjectReference)
-            {
-                return WellKnownSynchronizationKinds.ProjectReference;
-            }
-
-            if (value is MetadataReference)
-            {
-                return WellKnownSynchronizationKinds.MetadataReference;
-            }
-
-            if (value is AnalyzerReference)
-            {
-                return WellKnownSynchronizationKinds.AnalyzerReference;
-            }
-
-            if (value is TextDocumentState)
-            {
-                return WellKnownSynchronizationKinds.RecoverableSourceText;
-            }
-
-            if (value is SourceText)
-            {
-                return WellKnownSynchronizationKinds.SourceText;
-            }
-
-            if (value is OptionSet)
-            {
-                return WellKnownSynchronizationKinds.OptionSet;
+            switch (value)
```

Comment:
💭 I'd be curious how much time is spent doing type checks in this method, but the new behavior is similar to the old.